src/pcp/gadgets.rs

```diff
@@ -192,6 +192,176 @@ impl<F: FieldElement> Gadget<F> for PolyEval<F> {
+            bits,
+        }
+    }
+
+    pub(crate) fn call_poly_direct<V: AsRef<[F]>>(
+        &mut self,
+        outp: &mut [F],
+        inp: &[V],
+    ) -> Result<(), PcpError> {
+        let bits = self.bits;
+        let r_vec = &inp[..bits];
+        let x_vec = &inp[bits..2 * bits];
+        let x = inp[2 * bits].as_ref();
+
+        let z = poly_mul(x, x);
+        for i in 0..z.len() {
+            outp[i] = z[i];
+        }
+        for i in z.len()..outp.len() {
+            outp[i] = F::zero();
+        }
+
+        for l in 0..bits {
+            let mut z = r_vec[l].as_ref().to_vec();
+            for i in 0..3 {
+                for j in 0..z.len() {
+                    outp[j] += self.poly[i] * z[j];
+                }
+
+                if i < 2 {
+                    z = poly_mul(&z, x_vec[l].as_ref());
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub(crate) fn call_poly_fft<V: AsRef<[F]>>(
+        &mut self,
+        outp: &mut [F],
+        inp: &[V],
+    ) -> Result<(), PcpError> {
+        let bits = self.bits;
+        let n = self.n;
+        let r_vec = &inp[..bits];
+        let x_vec = &inp[bits..2 * bits];
+        let x = inp[2 * bits].as_ref();
+
+        let mut x_vals = vec![F::zero(); n];
+        let mut z_vals = vec![F::zero(); n];
+        let mut z = vec![F::zero(); n];
+
+        let m = n / 2;
+        let m_inv = self.n_inv * F::from(F::Integer::try_from(2).unwrap());
+        discrete_fourier_transform(&mut x_vals, x, m)?;
+        for j in 0..m {
+            z_vals[j] = x_vals[j] * x_vals[j];
+        }
+        discrete_fourier_transform(&mut z, &z_vals, m)?;
+        discrete_fourier_transform_inv_finish(&mut z, m, m_inv);
+        for i in 0..outp.len() {
+            outp[i] = z[i];
+        }
+
+        for l in 0..bits {
+            let x = x_vec[l].as_ref();
+            let y = r_vec[l].as_ref();
+            z[..y.len()].clone_from_slice(y);
+
+            discrete_fourier_transform(&mut x_vals, x, n)?;
+            discrete_fourier_transform(&mut z_vals, y, n)?;
+
+            let mut z_len = y.len();
+            for i in 0..3 {
+                for j in 0..z_len {
+                    outp[j] += self.poly[i] * z[j];
+                }
+
+                if i < 2 {
+                    for j in 0..n {
+                        z_vals[j] *= x_vals[j];
+                    }
+
+                    discrete_fourier_transform(&mut z, &z_vals, n)?;
+                    discrete_fourier_transform_inv_finish(&mut z, n, self.n_inv);
+                    z_len += x.len();
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl<F: FieldElement> GadgetCallOnly<F> for MeanVarUnsigned<F> {
+    fn call(&mut self, inp: &[F]) -> Result<F, PcpError> {
+        gadget_call_check(self, inp.len())?;
+        let bits = self.bits;
+        let r_vec = &inp[..bits];
+        let x_vec = &inp[bits..2 * bits];
+        let x = inp[2 * bits];
```

Comment:
This should check that `inp.len() == 2 * bits + 1`

Reply:
This is checked by `gadget_call_check` above: https://github.com/abetterinternet/libprio-rs/pull/48/files#diff-e01a59bccdbfaacea57d3cf97d4ad2fc0b9d8f1d333aeaa0098b150ad60a7a02R329.